File: Java/src/main/java/core/test1/Test1.java
package core.test1;

import java.io.BufferedReader;
import java.io.DataOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.Map;

import org.json.JSONArray;
import org.json.JSONObject;

import model.Triple;
import model.User;

public class Test1 {

	private static final String blazegraphURI = "http://localhost:9999/blazegraph/namespace/kb/sparql";
	private static final String graph = "http://test1/java";
	
	public static void init() throws IOException {
		//clean graph
		sendPOSTUpdate("DELETE WHERE {GRAPH <"+graph+"> { ?s ?p ?o }}");
	}
	
	public static String processRequest(Map<String,String> param) throws IOException, InterruptedException {
		String id = param.get("ID");
		if(id==null) {
			return "Error: ID is null";
		}
		User u = ServerSession.getUser(id);
		if(u==null) {
			u=new User(id);
			ServerSession.addUser(id,u);
		}
		//generate a random triple 
		Triple t = generateTriple(u.getMySubject());
		//save the generated triple in own server session
		u.add(t);
		//call Blazegraph to save it in the RDF-Store
		String resp = sendPOSTUpdate(t.generateInsertSPARQL(graph));
		if(resp==null) {
			throw new IOException("Post update resp is null");
		}

		//System.out.println(u.generateSparqlQuery(graph));
		//ask to blazegraph all the triples in the store, excluding their own triples
		resp=sendGETQuery(u.generateSparqlQuery(graph));
		if(resp==null) {
			throw new IOException("GET query resp is null");
		}
		System.out.println("\n\n---->"+resp.toString()+"\n");
		JSONObject obj = new JSONObject(resp);
		JSONArray arr = obj.getJSONObject("results").getJSONArray("bindings");
		//compare all the query result triples with all the triples in the own session, 
		//and if there is a match it will increase a counter
		String userStatus = "User "+id+" counter: "+ u.countMatch(arr);
		System.out.println(userStatus);
		return userStatus;
	}
	
	private static Triple generateTriple(String id) {
		String o = "obj_"+getRandomNumber(1,100);
		String p = "http://at_"+System.currentTimeMillis();
		return new Triple(id,p,o);
	}
	
	private static int getRandomNumber(int min, int max) {
	    return (int) ((Math.random() * (max - min)) + min);
	}
	
	
	private static String sendGETQuery(String query) throws IOException {
		String param = String.format("query=%s",  URLEncoder.encode(query, "UTF-8"));
		URL obj = new URL(blazegraphURI+"?"+param);
		//System.out.println(obj);
		HttpURLConnection con = (HttpURLConnection) obj.openConnection();
		con.setRequestMethod("GET");
		con.setRequestProperty("User-Agent", "Mozilla/5.0");
		con.setRequestProperty("Accept","application/json");
		int responseCode = con.getResponseCode();
		//System.out.println("GET Response Code :: " + responseCode);
		if (responseCode == HttpURLConnection.HTTP_OK) { // success
			BufferedReader in = new BufferedReader(new InputStreamReader(
					con.getInputStream()));
			String inputLine;
			StringBuffer response = new StringBuffer();

			while ((inputLine = in.readLine()) != null) {
				response.append(inputLine);
			}
			in.close();

			// print result
			return response.toString();
		} else {
			return null;
		}

	}
	
	private static String sendPOSTUpdate(String update) throws IOException {
		URL obj = new URL(blazegraphURI);
		HttpURLConnection con = (HttpURLConnection) obj.openConnection();
		con.setRequestProperty("Accept","application/json");
		String urlParameters  = "update="+update;
		byte[] postData       = urlParameters.getBytes( StandardCharsets.UTF_8 );
		int    postDataLength = postData.length;
		con.setDoOutput( true );
		con.setInstanceFollowRedirects( false );
		con.setRequestMethod( "POST" );
		con.setRequestProperty( "Content-Type", "application/x-www-form-urlencoded"); 
		con.setRequestProperty( "charset", "utf-8");
		con.setRequestProperty( "Content-Length", Integer.toString( postDataLength ));
		con.setUseCaches( false );
		try( DataOutputStream wr = new DataOutputStream( con.getOutputStream())) {
		   wr.write( postData );
		}
		int responseCode = con.getResponseCode();
		if (responseCode == HttpURLConnection.HTTP_OK) { 
			BufferedReader in = new BufferedReader(new InputStreamReader(
					con.getInputStream()));
			String inputLine;
			StringBuffer response = new StringBuffer();

			while ((inputLine = in.readLine()) != null) {
				response.append(inputLine);
			}
			in.close();
			return response.toString();
		} else {
			return null;
		}

	}
}

File: TypeScript/src/test1/Test1.ts
import axios from "axios";
import Triple from '../model/Triple';
import User from '../model/User';
import ServerSession from './ServerSession';
class Test1{
	private static blazegraphURI = "http://localhost:9999/blazegraph/namespace/kb/sparql";
	private static graph = "http://test1/typescript";

    public static init():void{
        Test1.sendPOSTUpdate("DELETE WHERE {GRAPH <"+Test1.graph+"> { ?s ?p ?o }}",()=>{
            console.log("RDF store clenned");
        },(err)=>{
           // console.log("Error Test1.sendPostUpdate: " + err);
            console.log("Post Error: " , err);
        });
    }
    
    public static processRequest(id:string,res):void{ 
		if(id===null || id===undefined) {
            // console.log("Error: ID is: ",id);
			res.send("Error: ID is null");
		}else{
            let u = ServerSession.getUser(id);
            if(u===null|| u===undefined) {
                u=new User(id);
                ServerSession.addUser(id,u);
            }
            //generate a random triple 
            const t = Test1.generateTriple(u.mySubject);
            //save the generated triple in own server session
            u.add(t);
            //call Blazegraph to save it in the RDF-Store
            //console.log("t.generateInsertSPARQL(Test1.graph)",t.generateInsertSPARQL(Test1.graph));
            Test1.sendPOSTUpdate(t.generateInsertSPARQL(Test1.graph),(ris)=>{
                if(ris===null || ris===undefined) {
                    // console.log("Post: ",ris);
                    res.send("Post update resp is null");
                }else{
                    //ask to blazegraph all the triples in the store, excluding their own triples
                    Test1.sendGETQuery(u.generateSparqlQuery(Test1.graph),(ris2)=>{
                        if(ris2===null || ris2?.data?.results?.bindings===undefined) {
                            // console.log("ris2: ",ris2);
                            res.send("GET query resp is null");
                        }else{
                            //compare all the query result triples with all the triples in the own session, 
                            //and if there is a match it will increase a counter
                            const userStatus = "User "+id+" counter: "+ u.countMatch(ris2.data.results.bindings);
                            console.log(userStatus);
                            res.send(userStatus);
                        }
                    },(err2)=>{
                        res.send("GET Error: "+ err2);
                    });
                 
                }
            },(err)=>{
                // console.log("Post Error: ",err);
                res.send("Post Error: "+ err);
            });
            
        }

    }

	private static generateTriple(id:string):Triple {
		const o = "obj_"+Test1.getRandomNumber(1,100);
		const p= "http://at_"+new Date().getTime();
		return new Triple(id,p,o);
	}
	
	private static getRandomNumber(min:number, max:number):number {
        return Math.trunc((Math.random() * (max - min)) + min);
	}
    private static sendPOSTUpdate(sparqlUpdate:string,cbok:(ris) => void,cbErr:(err) => void){    
        const config = {
            headers: {
                'Accept': 'application/json'
            }
        }   
        const params = new URLSearchParams();
        params.append('update', sparqlUpdate);
        axios.post(Test1.blazegraphURI,params,config)
        .then(cbok)
        .catch(cbErr);
    }

    private static sendGETQuery(sparqlQuery:string,cbok:(ris) => void,cbErr:(err) => void){   
        const config = {
            headers: {
                Accept: 'application/json',
            }
        }          
        axios.get(Test1.blazegraphURI+"?query="+sparqlQuery,config)
        .then(cbok)
        .catch(cbErr);
    }
}

export default Test1;